pkg/provisioner/ironic/ironic.go

```diff
@@ -609,6 +609,59 @@ func (p *ironicProvisioner) UpdateHardwareState() (result provisioner.Result, er
 	return result, nil
 }
 
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	var cleanSteps []nodes.CleanStep
+
+	if p.bmcAccess.RAIDInterface() != "" {
+		// Set raid config
+		err = setTargetRAIDCfg(p.client, ironicNode, p.host.Status.Provisioning.RAID)
+		if err != nil {
+			p.log.Error(err, "failed in setting the target RAID configuration")
+			return
+		}
+		// Build raid configuration clean steps
+		cleanSteps = append(cleanSteps, buildRAIDCleanSteps(p.host.Status.Provisioning.RAID)...)
+	} else {
+		p.log.Info("the BMC of the current node does not support RAID", "BMC", p.bmcAccess.Driver())
+	}
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		var success bool
+		success, result, err = p.tryChangeNodeProvisionState(ironicNode, opts)
+		result.Dirty = !success
```

Comment:
This is surprising. I would expect the result to be marked as dirty if the node is successfully put into the clean state. Why is the logic reversed?

Reply:
Because the state of `ironic` before and after `manual cleaning` is `manageable`, we do `manual clean` in `preparing` state, but check `manual clean`'s result in `provisioning` state. So if we do `manual clean` success, we need let state go to `provisioning` to check the result of `manual clean`.

Reply:
The `Dirty` flag needs to communicate at least 2 different things: whether or not there are changes to the Host to write, and whether or not the operation is complete. These are orthogonal, and we only get 1 bit to communicate them in. In this case we're using Dirty = false to communicate that the operation is complete. Some day the Provisioner API needs an overhaul, but for now this is correct.